orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactUtils.java

```diff
@@ -228,65 +225,44 @@ public ArtifactUtils(
-                        .collect(toList()))
-            .orElse(emptyList());
-
-    List<Artifact> receivedArtifactsFromPipeline =
-        Optional.ofNullable((List<?>) pipeline.get("receivedArtifacts"))
-            .map(
-                list ->
-                    list.stream()
-                        .map(it -> objectMapper.convertValue(it, Artifact.class))
-                        .collect(toList()))
-            .orElse(emptyList());
-    List<Artifact> artifactsFromTrigger =
-        Optional.ofNullable((List<?>) trigger.get("artifacts"))
-            .map(
-                list ->
-                    list.stream()
-                        .map(it -> objectMapper.convertValue(it, Artifact.class))
-                        .collect(toList()))
-            .orElse(emptyList());
-
-    List<Artifact> receivedArtifacts =
-        Stream.concat(receivedArtifactsFromPipeline.stream(), artifactsFromTrigger.stream())
-            .distinct()
-            .collect(toList());
-
-    if (expectedArtifacts.isEmpty()) {
-      try {
-        trigger.put(
-            "artifacts",
-            objectMapper.readValue(objectMapper.writeValueAsString(receivedArtifacts), List.class));
-      } catch (IOException e) {
-        log.warn("Failure storing received artifacts: {}", e.getMessage(), e);
-      }
-      return;
-    }
-
-    List<Artifact> priorArtifacts = getPriorArtifacts(pipeline);
-    LinkedHashSet<Artifact> resolvedArtifacts =
-        resolveExpectedArtifacts(expectedArtifacts, receivedArtifacts, priorArtifacts, true);
-    LinkedHashSet<Artifact> allArtifacts = new LinkedHashSet<>(receivedArtifacts);
-    allArtifacts.addAll(resolvedArtifacts);
+    ImmutableList<ExpectedArtifact> expectedArtifacts =
+        Optional.ofNullable((List<?>) pipeline.get("expectedArtifacts")).orElse(emptyList())
+            .stream()
+            .map(it -> objectMapper.convertValue(it, ExpectedArtifact.class))
+            .collect(toImmutableList());
+
+    ImmutableSet<Artifact> receivedArtifacts =
+        Stream.concat(
+                Optional.ofNullable((List<?>) pipeline.get("receivedArtifacts")).orElse(emptyList())
+                    .stream(),
+                Optional.ofNullable((List<?>) trigger.get("artifacts")).orElse(emptyList())
+                    .stream())
+            .map(it -> objectMapper.convertValue(it, Artifact.class))
+            .collect(toImmutableSet());
+
+    ArtifactResolver.ResolveResult resolveResult =
+        ArtifactResolver.getInstance(receivedArtifacts, () -> getPriorArtifacts(pipeline), true)
+            .resolveExpectedArtifacts(expectedArtifacts);
+
+    ImmutableSet<Artifact> allArtifacts =
+        ImmutableSet.<Artifact>builder()
+            .addAll(receivedArtifacts)
+            .addAll(resolveResult.getResolvedArtifacts())
+            .build();
 
     try {
       trigger.put(
           "artifacts",
           objectMapper.readValue(objectMapper.writeValueAsString(allArtifacts), List.class));
```

Comment:
This is a bit of a tangent to the exciting change you're making here, so feel free to ignore.

I was in this neighborhood a couple of weeks ago trying to debug a situation where spinnaker was picking up an artifact from a prior execution, even though I thought I specified enough via manual execution.  I'm still not quite sure what was happening, though I was able to disable use prior artifact in those pipelines, so if a similar thing happens again I figure I'll get a failed pipeline execution instead of using the 'wrong' artifact.

Anyway, the struggle I had was that by changing trigger.artifacts here, it's hard to see what actually came via the trigger vs. what happened during artifact resolution.  This would help me figure out whether my struggle was in e.g. deck or gate vs. here.

A log message would help, but if there's a way to reflect this in the pipeline execution json, it'd facilitate my troubleshooting even more.

Thanks for reading :)

Reply:
I agree this is confusing...in general I think it's confusing that sometimes we store artifacts in `pipeline.receivedArtifacts` and sometimes in `trigger.artifacts` (and in this case concatenate the two and store them on the trigger).  One day it would be great to refactor this and have a consistent contract for where artifacts should go to be resolved...but I'm worried about making any changes to that here as I'm sure there are a lot of things depending on it.

Reply:
Yeah, had a feeling changes here would ripple out a bunch.  Would it help if I create a separate issue?

Reply:
Yeah, that would be helpful---I can't promise any changes to how this works in the near future as it might be a big project, but I'd definitely like to understand better the confusing case you ran into.

Reply:
Here goes: https://github.com/spinnaker/spinnaker/issues/5344